homeassistant/components/climate/dyson.py

```diff
@@ -0,0 +1,194 @@
+        return SUPPORT_FLAGS
+
+    @property
+    def name(self):
+        """Return the display name of this climate."""
+        return self._device.name
+
+    @property
+    def temperature_unit(self):
+        """Return the unit of measurement."""
+        return TEMP_CELSIUS
+
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        if self._device.environmental_state:
+            temperature_kelvin = self._device.environmental_state.temperature
+            if temperature_kelvin != 0:
+                self._current_temp = float("{0:.1f}".format(
+                    temperature_kelvin - 273))
+        return self._current_temp
+
+    @property
+    def target_temperature(self):
+        """Return the target temperature."""
+        if self._pending_target_temp is None:
+            heat_target = int(self._device.state.heat_target) / 10
+            return int(heat_target - 273)
+        return self._pending_target_temp
+
+    @property
+    def current_humidity(self):
+        """Return the current humidity."""
+        if self._device.environmental_state:
+            if self._device.environmental_state.humidity == 0:
+                return None
+            return self._device.environmental_state.humidity
+        return None
+
+    @property
+    def current_operation(self):
+        """Return current operation ie. heat, cool, idle."""
+        from libpurecoollink.const import HeatMode, HeatState
+        if self._device.state.heat_mode == HeatMode.HEAT_ON.value:
+            if self._device.state.heat_state == HeatState.HEAT_STATE_ON.value:
+                return STATE_HEAT
+            return STATE_IDLE
+        return STATE_COOL
+
+    @property
+    def operation_list(self):
+        """Return the list of available operation modes."""
+        return OPERATION_LIST
+
+    @property
+    def current_fan_mode(self):
+        """Return the fan setting."""
+        from libpurecoollink.const import FocusMode
+        if self._device.state.focus_mode == FocusMode.FOCUS_ON.value:
+            return STATE_FOCUS
+        return STATE_DIFFUSE
+
+    @property
+    def fan_list(self):
+        """Return the list of available fan modes."""
+        return FAN_LIST
+
+    def set_temperature(self, **kwargs):
+        """Set new target temperature."""
+        target_temp = kwargs.get(ATTR_TEMPERATURE)
+        if target_temp is None:
+            return
+        target_temp = int(target_temp)
+        _LOGGER.debug("Set %s temperature %s", self.name, target_temp)
+        # Limit the target temperature into acceptable range.
+        target_temp = min(self.max_temp, target_temp)
+        target_temp = max(self.min_temp, target_temp)
+        self._pending_target_temp = target_temp
+        # Dyson only response when it is in heat mode.
+        from libpurecoollink.const import HeatMode
+        if self._device.state.heat_mode == HeatMode.HEAT_ON.value:
```

Comment:
We should first check this condition, and failed method fast

Reply:
I do not understand how to fail this method faster? Assuming you agree with the pending value method, I will still need to do all the checking and store the value regardlessly.

Reply:
```python
    def set_temperature(self, **kwargs):
        """Set new target temperature."""
        target_temp = kwargs.get(ATTR_TEMPERATURE)
        if target_temp is None:
            return

        # Dyson only response when it is in heat mode.
        from libpurecoollink.const import HeatMode
        if self._device.state.heat_mode != HeatMode.HEAT_ON.value:
            _LOGGER.error("Cannot set target temperature, bla bla")
            return

        target_temp = int(target_temp)
        # Limit the target temperature into acceptable range.
        target_temp = min(self.max_temp, target_temp)
        target_temp = max(self.min_temp, target_temp)
        _LOGGER.debug("Set %s temperature %s", self.name, target_temp)
        self._set_temperature_on_device(target_temp)
```

Reply:
I think with this approach, it will really confuse the user in what they expect (as with I did when I first started developing this component.)
The limitations of this approach are:
1. The discrepancy from the physical interface.
If you look at the physical remote:
![remote](https://images-na.ssl-images-amazon.com/images/I/51W-44npbTL._SX425_.jpg)
...there really isn't an **on/off** switch for cool/hot mode. The top right button always switches to cool mode, and there are temperature control buttons for up and down. Those two buttons control the temperature, but also always will affect the target temperature regardless of heat/cool mode. In fact, it switches to heat mode if it was previously in cool mode. (Internally, I think *it makes a call to switch to heat mode if it is in cool mode, then makes a call to increase/decrease temperature*; which is different than our current approach of controlling the heat mode on/off and temp separately. This leads to my approach 3 later on).
2. Within the front end interface, changing the target temp via up/down arrowhead instantly changes the temperature value (even though its state remains the same). It would be out of sync (and not what the user expects) where the user would think they successfully changed the temperature but in reality, the request was simply ignored.
 * Logging an error doesn't really make sense because the upstream lib nor device will generate an error or any messages. It just simply has no effect.

I have given a thought when I first developed this component. There are possibly 3 approaches to this:
1. Simple and naively just send the request without any checking. It was how I did it when I first started, but it really confuses me due to the inconsistent interface from its homeassistant and physical representation. And it renders the user input useless (without a concrete way of knowing the request was ignored because the updates of device states need to wait for the mqtt callback)
2. My current approach, which simply stores the value for now, and re-deploy later on. This keep track of user wanted temp value and apply it when appropriate. Whether it's best or not it is debatable, but it feels most natural to me.
3. Simply mimic the physical interface where we check if it is in heat mode when user changes target temp. If it is not, then we first make a call to switch heatmode on, then make a call to change the target temp (Technically we can combine both calls to one because mqtt request can set multiple configurations together). This will result in changing temperature will switch heat mode on (which is, in fact, **the most consistent** with the physical interface as it is how Dyson does it in its physical remote and Android/iPhone app). However, it is a bit different from how normal homeassistant climate device works I assume?

Please let me know which you think is more appropriate thanks.

Reply:
@awarecan or @balloob please give you're final opinion so we can move this PR along.

Reply:
> Within the front end interface, changing the target temp via up/down arrowhead instantly changes the temperature value (even though its state remains the same). It would be out of sync (and not what the user expects) where the user would think they successfully changed the temperature but in reality, the request was simply ignored.

The frontend will wait a few seconds before calling set temperature. After setting, it will reset the frontend to the current state of the climate. So it should be clear to the user that the operation was not successful.

Reply:
Sure!